### PR TITLE
Add pull request template

### PR DESCRIPTION
**Pull request by mosabua**
As agreed with @martint here is a first draft. This is aimed to improve release process and contributions to be scoped better, and to make the relevant information more easily available. 

Open to any feedback for improvements/adjustments.

### CODE REVIEW

File: .github/pull_request_template.md
Comment (dain) [inline .github/pull_request_template.md:10]: Maybe prefill with "none" or add a comment saying "Remove this section if not needed"

Comment (mosabua) [reply .github/pull_request_template.md]: done

File: .github/pull_request_template.md
Comment (mosabua) [inline .github/pull_request_template.md:29]: This will basically NOT render in the display but help when editing.. we can do that for others as well.

File: .github/pull_request_template.md
Comment (findepi) [inline .github/pull_request_template.md:24]: That should go into the PR title.

File: .github/pull_request_template.md
Comment (findepi) [inline .github/pull_request_template.md:22]: how does it help?

File: .github/pull_request_template.md
Comment (findepi) [inline .github/pull_request_template.md:14]: Can we remove `docs-needed` label now?

File: .github/pull_request_template.md
Comment (findepi) [inline .github/pull_request_template.md:6]: Encourage use of "Fixes ..." when PR supposed to close an issue.

File: .github/pull_request_template.md
Comment (findepi) [inline .github/pull_request_template.md:10]: This is counter-productive. people will feel obliged to find *some* related PRs, which doesn't really help.

Comment (hashhar) [reply .github/pull_request_template.md:10]: It also seems to encourage creating separate doc prs which is counterproductive as well IMO.

File: .github/pull_request_template.md
Comment (mosabua) [reply .github/pull_request_template.md:24]: No .. author should be explicit .. e.g. if something affects base jdbc class and some but not all jdbc based connectors.. the contributor and maintainer should assemble that list and add it here, so it is ready for the release notes assembly..

File: .github/pull_request_template.md
Comment (mosabua) [reply .github/pull_request_template.md:22]: It forces the contributor and merging maintainer to think through release notes entry from the start and latest at merge. And it creates a standard place with the PR to have that info ..

File: .github/pull_request_template.md
Comment (mosabua) [reply .github/pull_request_template.md:14]: Not sure yet .. we are just assembling more info in the PR description for now .. labels can help us as well. We are figuring this out. Labels are not consistently used at this stage

File: .github/pull_request_template.md
Comment (mosabua) [reply .github/pull_request_template.md:6]: Should I add a comment for that or a into the template directly

```
* Fixes #issuenumber
```

File: .github/pull_request_template.md
Comment (mosabua) [reply .github/pull_request_template.md:10]: It is currently not realistic to only support code and doc PRs together. It is more the exception than the rule. Even maintainers, but esp. many contributors, rely on the help from others to get the docs done. 

Also ... the default is none .. so nobody should feel like they should find a random one to link ..

File: .github/pull_request_template.md
Comment (mosabua) [reply .github/pull_request_template.md:24]: It is also not realistic if more than two connectors are affected..

Comment (findepi) [reply .github/pull_request_template.md:24]: so the short version of this should go into title (one connector - easy, many JDBC connectors -- easy, just say "jdbc connectors")
the longer and more precise version would go at the bottom, where rn entry is suggested. awesome, so this point is redundant

remove

File: .github/pull_request_template.md
Comment (findepi) [reply .github/pull_request_template.md:22]: this looks as written by a release notes maniac, as release notes are being now mentioned in 3 different places.
the redundancy is going to irritate, not help

File: .github/pull_request_template.md
Comment (findepi) [reply .github/pull_request_template.md:6]: `Fixes #issuenumber` -- like that. Doesn't require a bullet.
And add some comment to remove this line if there is no issue.

File: .github/pull_request_template.md
Comment (findepi) [reply .github/pull_request_template.md:10]: > It also seems to encourage creating separate doc prs which is counterproductive as well IMO.

@hashhar depends on the change. for small changes together is better, for bigger changes separate is better

@mosabua i don't see a need for `Related code and/or documentation pull requests` section, please remove

File: .github/pull_request_template.md
Comment (mosabua) [reply .github/pull_request_template.md]: Updated as discussed with @electrum

File: .github/pull_request_template.md
Comment (mosabua) [reply .github/pull_request_template.md]: Updated as discussed with @electrum

File: .github/pull_request_template.md
Comment (mosabua) [reply .github/pull_request_template.md]: Updated as discussed with @electrum

File: .github/pull_request_template.md
Comment (hashhar) [inline .github/pull_request_template.md:40]: This puts lot of burden on PR submitter to provide release notes entry. Maintainers can always edit the PR description anyway and add them there if we all are agreed that we'll collect release notes from the PR description rather than the "Release notes for XXX" issue we create today.

More useful content here would be to link to https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#commits-and-pull-requests maybe?

Comment (findepi) [reply .github/pull_request_template.md:40]: > This puts lot of burden on PR submitter to provide release notes entry. Maintainers can always edit the PR description anyway and add them there if we all are agreed that we'll collect release notes from the PR description rather than the "Release notes for XXX" issue we create today.

i agree that this move of responsibility isn't what we want.

Comment (mosabua) [reply .github/pull_request_template.md:40]: I wasnt aware that maintainers can edit the description.. that make it easier. I will reword and add the link

File: .github/pull_request_template.md
Comment (findepi) [inline .github/pull_request_template.md:44]: The PR template doesn't (and shouldn't) contain RN template. 

Also, we currently collect RNs under a RN issue.
I don't think anything changed, and there even was no Maintainers' dicussion advocating for such a change.

**Comment by findepi** [comment]
Instead of introducing a bloated and controversial PR template, can we have a lean one, that we agree about? Iterate.

File: .github/pull_request_template.md
Comment (martint) [reply .github/pull_request_template.md:44]: > I don't think anything changed, and there even was no Maintainers' dicussion advocating for such a change.

We've had some ad-hoc discussions and hence the PR. We've talked about possibly doing something like this many months ago (how we might streamline the release note collection process), and @mosabua now gracefully offered to help try some ideas and drive this forward.

**Comment by mosabua** [comment]
Here is a live PR that used the template already manually - https://github.com/trinodb/trino/pull/10898

File: .github/pull_request_template.md
Comment (hashhar) [reply .github/pull_request_template.md:44]: Collecting RNs under PR doesn't scale well IMO with external contributors - now non-PR related maintainanence discussion ends up happening during PR review adding noise.
Also it's easy to end up with outdated release notes since the actual changes being merged is not final until the PR is actually merged.

I'd propose to exclude the RN template for now - we can iterate on this over time.

**Comment by findepi** [comment]
> Here is a live PR that used the template already manually - #10898

i read there

> * Update JDBC driver to 10.2.0 ({issue}`10898`)


This is a good example why we shouldn't ask contributors for release notes.

in that PR, a contributor didn't now the release notes need to present the change from user perspective.
Version upgrade of an underlying library hardly translates to user visible impact, unless user does their job and checks why we could have done the upgrade.

File: .github/pull_request_template.md
Comment (mosabua) [reply .github/pull_request_template.md:44]: I strongly disagree. A proper RN entry and documentation should be part of the work of getting a PR ready for merge and everything should be done by the time something is merged. This work can and should be done by the maintainer helping the contributor to get to merge. And it has to explicitly communicate if a RN or docs are needed for each PR. 

> > Here is a live PR that used the template already manually - #10898
> 
> i read there
> 
> > * Update JDBC driver to 10.2.0 ({issue}`10898`)
> 
> This is a good example why we shouldn't ask contributors for release notes.
> 
> in that PR, a contributor didn't now the release notes need to present the change from user perspective. Version upgrade of an underlying library hardly translates to user visible impact, unless user does their job and checks why we could have done the upgrade.

If you actually looked at the PR in detail you would see that this is WIP and only a draft proposal. Part of getting the PR ready is to figure out the correct release notes entry and the PR says as much. There are most likely user facing changes and they are in the linked upstream release notes. By the time this PR is ready it should include a release notes entry that is ready to be used nearly unmodified. Since this is WIP that is of course not the case yet..

File: .github/pull_request_template.md
Comment (dain) [inline .github/pull_request_template.md:22]: This section is pretty confusing to me.  I believe the goal here is to have a "documentation todo" list, but both are default to checked.
I suggest we rename this section to "Documentation TODOs" or "Review TODOs", so we can add other items for reviewers to check off.  Also, I think they should start unchecked, and be reworded to sound like TODOs.  For example:
```
- [ ] Assess need for user facing documentation
- [ ] Add necessary release notes
```

**Comment by mosabua** [comment]
Everyone please keep in mind that this is all just a template .. there is no enforcement to use it.

File: .github/pull_request_template.md
Comment (mosabua) [reply .github/pull_request_template.md:22]: Ideally we get the following info:

* Release notes entry needed? either confirm it is not needed or proposal for text
* Docs affected and needed and then info on if docs is in PR or in separate PR or there is at least a issue to get the docs done

Any ideas how we get that into the template better?

File: .github/pull_request_template.md
Comment (hashhar) [reply .github/pull_request_template.md:44]: >  Also it's easy to end up with outdated release notes since the actual changes being merged is not final until the PR is actually merged.

And there is no mechanism to indicate that the RN is no longer a draft or that someone forgot to update it.

File: .github/pull_request_template.md
Comment (dain) [reply .github/pull_request_template.md:22]: For the release notes, the maintainer would edit the description and add them inline.  For the docs, what do you think the maintainer would do?  At the very least, I'd like to see a section where the maintainer says, "we're going to need docs to cover x and y", but ideally there would be changes to the docs in the PR.

Comment (hashhar) [reply .github/pull_request_template.md:22]: What about instead of check-boxes we (maintainers) start using labels `docs-needed` and `release-notes-needed` to mark PRs that need a RN/docs.

It has multiple benefits, it makes it possible to automate the creation of the list of commits + whether RN or docs are needed or not and it makes it easy to search for such PRs.

File: .github/pull_request_template.md
Comment (mosabua) [reply .github/pull_request_template.md:44]: Sure .. everything in the PR is valid until merge. Then the release notes text is absorbed into the release notes PR .. and iteration continues there if necessary.

File: .github/pull_request_template.md
Comment (mosabua) [reply .github/pull_request_template.md:22]: If we somehow make that consistently happen .. sure.. its just easier for contributors this way since they cant assign labels..

Comment (mosabua) [reply .github/pull_request_template.md:22]: I rejigged it along the idea of @dain to have a kind of questionnaire to answer .

**Review by dain — APPROVED** [review]
I think this is good enough to try out.  If we don't like it, we can always remove it.

**Comment by mosabua** [comment]
@dain and myself found some issues with the checkbox syntax interfering with GHA .. should be all fixed now

**Review by martint — APPROVED** [review]
Agreed. This is a good start! Once we see how people end up using it in practice, we can adjust it.

**Pull request merged**
